src/cpp/session/modules/clang/CodeCompletion.cpp

```diff
@@ -154,6 +169,237 @@ core::json::Object toJson(const CodeCompleteResult& result)
+   completionJson["type"]       = completionType;
+   completionJson["typed_text"] = name;
+   
+   json::Array textJson;
+   json::Object objectJson;
+   objectJson["text"]    = name;
+   objectJson["comment"] = source;
+   textJson.push_back(objectJson);
+   completionJson["text"] = textJson;
+
+   return completionJson;
+}
+
+Error getSystemHeaderCompletions(const std::string& token,
+                                 const std::string& parentDir,
+                                 const FilePath& filePath,
+                                 const std::string& docId,
+                                 const core::json::JsonRpcRequest& request,
+                                 core::json::JsonRpcResponse* pResponse)
+{
+   std::vector<std::string> includePaths;
+   
+   // discover the system headers
+   discoverSystemIncludePaths(&includePaths);
+   
+   // discover TU-related include paths
+   discoverTranslationUnitIncludePaths(filePath, &includePaths);
+   
+   // remove dupes
+   std::sort(includePaths.begin(), includePaths.end());
+   includePaths.erase(std::unique(includePaths.begin(), includePaths.end()),
+                      includePaths.end());
+   
+   // loop through header include paths and return paths that match token
+   std::set<std::string> discoveredEntries;
+   json::Array completionsJson;
+   
+   BOOST_FOREACH(const std::string& path, includePaths)
+   {
+      FilePath includePath(path);
+      if (!includePath.exists())
+         continue;
+      
+      FilePath targetPath = includePath.complete(parentDir);
+      if (!targetPath.exists())
+         continue;
+      
+      std::vector<FilePath> children;
+      Error error = targetPath.children(&children);
+      if (error)
+         LOG_ERROR(error);
+      
+      BOOST_FOREACH(const FilePath& childPath, children)
+      {
+         std::string name = childPath.filename();
+         if (discoveredEntries.count(name))
+            continue;
+         
+         if (string_utils::isSubsequence(name, token, true))
+         {
+            int type = childPath.isDirectory() ? kCompletionDirectory : kCompletionFile;
+            completionsJson.push_back(jsonHeaderCompletionResult(name,
+                                                                 childPath.absolutePath(),
+                                                                 type));
+         }
+         
+         discoveredEntries.insert(name);
+      }
+   }
+   
+   // construct completion result
+   json::Object resultJson;
+   resultJson["completions"] = completionsJson;
+   pResponse->setResult(resultJson);
+   return Success();
+}
+
+Error getLocalHeaderCompletions(const std::string& token,
+                                const std::string& parentDir,
+                                const FilePath& filePath,
+                                const core::json::JsonRpcRequest& request,
+                                core::json::JsonRpcResponse* pResponse)
+{
+   // TODO -- list files + dirs
```

Comment:
This is the stub for enabling autocompletion of plain paths in e.g.

    #include "foo/|

Let me see if this is easy enough to implement.